# Исследование надежности заемщиков


**Описание проекта**

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.
Результаты исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.
Описание данных

- `children` — количество детей в семье
- `days_employed` — общий трудовой стаж в днях
- `dob_years` — возраст клиента в годах
- `education` — уровень образования клиента
- `education_id` — идентификатор уровня образования
- `family_status` — семейное положение
- `family_status_id` — идентификатор семейного положения
- `gender` — пол клиента
- `income_type` — тип занятости
- `debt` — имел ли задолженность по возврату кредитов
- `total_income` — ежемесячный доход
- `purpose` — цель получения кредита

Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

data = pd.read_csv('C:/Users/Sam/DA/Projects_YandexPracticum/Project2_bank_credit/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5], dtype=int64)

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5], dtype=int64)

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

Для определения зависимости между количество детей и возвратом кредита в срок создадим сводную таблицу. Сгруппируем информацию по количеству детей и возврату кредита, далее посчитаем процент невозвратных кредитов в каждой группе.

In [22]:
# методом sum узнаем общее число невозвратных кредитов, а методом count - общее число кредитов.
children_and_debt = data.pivot_table(index='children', values='debt', aggfunc=['sum', 'count'])
display(children_and_debt)
#children_and_debt.columns

sum  count
          debt   debt
children             
0         1063  14091
1          444   4808
2          194   2052
3           27    330
4            4     41
5            0      9

In [23]:
# вычислим процент невозвратных кредитов
children_and_debt['percent_debt'] = children_and_debt['sum']['debt'] / children_and_debt['count']['debt'] * 100
display(children_and_debt)

sum  count percent_debt
          debt   debt             
children                          
0         1063  14091     7.543822
1          444   4808     9.234609
2          194   2052     9.454191
3           27    330     8.181818
4            4     41     9.756098
5            0      9     0.000000

**Вывод:** 

После группировки данных сделаем выводы.
Клиент без детей имеет задолженность по кредиту в 7.54%. Наличие 1-4 детей у клиента увеличивает процент невозврата кредита в пределах от 0.6 до 2.2%. Показатель, что клиенты с 5 детьми не имеют задолженности по кредитам, отбросим из-за малой выборки (всего 9 человек).

**Итоговый вывод:** Есть небольшая зависимость между количеством детей и возвратом кредита в срок. Наличие детей немного увеличивает процент невозврата кредита на 2.2%.

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

Теперь выясним если ли зависимость между семейным положением и возвратом кредита в срок. Сгруппируем сводную таблицу по статус семейного положения и возврату кредита.

In [24]:
# методом mean узнаем долю невозвратных кредитов, debt принимает значения 1 и 0, т.е. кол-во невозвратных кредитов от общего числа
family_and_debt = data.pivot_table(index='family_status', values='debt', aggfunc='mean')
display(family_and_debt)

debt
family_status                  
Не женат / не замужем  0.097639
в разводе              0.070648
вдовец / вдова         0.066246
гражданский брак       0.093130
женат / замужем        0.075606

**Вывод:** 

Мы видим, что группы `гражданский брак` и `Не женат / не замужем` имеют самые большие уровни невозврата кредита в 9.3% и 9.8%.
С другой стороны, группа `вдовец / вдова` имеет самый низкий уровень невозврата кредита: 6.6%.

В группах `в разводе` и `женат / замужем` процент невозврата уже повыше, на уровне 7.1% и 7.6%.

**Итоговый вывод:** Мы видим некоторую зависимость от семейного статуса. Самые большие уровни невозврата показывают люди, которые не состоят в официальном браке, это группы `гражданский брак` и `Не женат / не замужем`. При этом разница в возврате кредита между этими группами и группой `вдовец / вдова`, у которой самый высокий возврат кредита, составляет всего 3.1%

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

Теперь определим зависимость между уровнем дохода и возвратом кредита в срок.

In [25]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.
data.head(1)
# методом mean узнаем долю невозвратных кредитов
income_and_debt = data.pivot_table(index=['total_income_category'], values='debt', aggfunc='mean')
display(income_and_debt)

debt
total_income_category          
A                      0.080000
B                      0.070602
C                      0.084982
D                      0.060172
E                      0.090909

Вспомним, как проходила категоризация по уровню дохода:

- 0–30000 — 'E';

- 30001–50000 — 'D';

- 50001–200000 — 'C';

- 200001–1000000 — 'B';

- 1000001 и выше — 'A'.

**Вывод:** 

Итак, видна зависимость между уровнем дохода и возвратом кредита. Люди с доходом до 30000 больше всего не возвращают кредиты. Наоборот, люди с доходом от 30000 до 50000 имеют самый высокий уровень возврата кредита. Люди с доходом от 50000 имеют средний уровень возврата кредита.

**Итоговый вывод:** есть зависимость между уровнем дохода и возвращаемостью кредита. Логично, люди с низким доходом имеют самый низкий процент возврата кредита в 9.1%. При этом люди с доходом чуть выше имеют самый высокий уровень возврата кредита в 6%. Люди со средним и выше среднего доходом и имеют средний уровень возврата кредита 7.1% - 8.5%.

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

Далее посмотрим наличие зависимости между целями кредита и его возвратом.

In [26]:
# применим метод mean для подсчета доли невозвратных кредитов
purpose_and_debt = data.pivot_table(index='purpose_category', values='debt', aggfunc='mean')
display(purpose_and_debt)

debt
purpose_category                  
операции с автомобилем    0.093480
операции с недвижимостью  0.072551
получение образования     0.092528
проведение свадьбы        0.079118

**Вывод:** 

Мы видим, что кредиты на автомобиль и образование являются самыми невозвратными на уровне 9.3%.
С другой стороны, кредиты на недвижимость и свадьбу являются наиболее возвращаемыми на уровне 7.3% и 7.9%.

**Итоговый вывод:** Есть некоторая зависимость между целями на кредит. Самые невозвратные кредиты на автомобиль и образование, а самые возвращаемые кредиты на недвижимость и свадьбу. Но разница между этими группами всего 2%.

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* 

Причины пропусков в данных бывают техническими, либо связаны с человеческим фактором. 
В данном датафрейме пропуски встречаются в столбцах `days_employed` и `total_income`.

Предположим, что люди могли пропустить ввод этих данных, либо могли ввести данные неправильно. Поэтому данные не были сохранены.

Также датафрейм мог быть выгружен с ошибками, либо данные были повреждены во время хранения.

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* 

Пропуски в количественных переменных заполняют характерным значением. Для заполнения выбирают значения, которые наиболее подходят к данным. Это либо медианное значение, либо среднее значение. 

Но среднее значение некорректно будет представлять выборку, если в ней есть очень большие или очень маленькие значения. В этом случае значение сдвинется от объективного. 

А медианное значение разделяет выборку ровно пополам, что является более оптимальным.

### Шаг 4: общий вывод.

Напишите ваш общий вывод.

Мы провели исследование клиентов банка на предмет их платежеспособности. Датафрейм был обработан, мы заполнили пропуски, удалили дубликаты и обработали аномальные значени. После подготовки данных, мы ответили на поставленные вопросы исследования. 

**Выводы:**

1. Есть ли зависимость между количеством детей и возвратом кредита в срок?

Есть небольшая зависимость между количеством детей. Люди без детей имеют самых высокий уровень возврата кредита. При этом наличие детей немного увеличивает процент невозврата кредита на 2.2%.


2. Есть ли зависимость между семейным положением и возвратом кредита в срок?

Есть небольшая зависимость от семейного статуса. Самые большие уровни невозврата показывают люди, которые не состоят в официальном браке, это группы `гражданский брак` и `Не женат / не замужем`. При этом разница в возврате кредита между этими группами и группой `вдовец / вдова`, у которой самый высокий возврат кредита, составляет всего 3.1%


3. Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

Есть зависимость между уровнем дохода и возвращаемостью кредита. Люди с низким доходом имеют самый низкий процент возврата кредита. При этом люди с доходом чуть выше имеют самый высокий уровень возврата кредита. Разница между этими группами составляет всего 3.1%.


4. Как разные цели кредита влияют на его возврат в срок?

Есть некоторая зависимость между целями на кредит. Самые невозвратные кредиты на автомобиль и образование, а самые возвращаемые кредиты на недвижимость и свадьбу. Но разница между этими группами всего 2%.
 
 
**Рекомендации:**

1. Необходимо узнать какой уровень невозврата кредита считается в банке критичным, т.к. в ходе исследования разница между группами с низким возвратом и высокии возвратом кредита была всего 3%.

2. Стоит уточнить у разработчиков базы данных банка о причинах пропусков в датафрейме и наличию отрицательных значений. Возможно стоит изменить запись, хранение и выгрузку, чтобы избежать данных проблем.

3. Для уменьшения вариантов целей на кредит в обрабатываемом датафрейме стоит ввести фиксированный список целей на кредит. При записи данных нужно будет выбрать один из вариантов.